[WIP] 26 update is corrected field

itshesam opened the pull request:
@danielzeljko 

closes #26 

I added ```seed.sql``` as fixtures to make it easier to write tests.
Also added ```vscode``` directory to gitignore, it was annoying for me :D

If number of correctors get 0 then the ```is_corrected``` field should be 0, right ?
And added the management command as you said, but i don't know how much load is on the database and don't have any estimation of the number of posts that need to be updated, so i just wrote the simplest possible way.

There's something problem in japanese splitting, it makes the pipline fail, i didn't find how to fix it, maybe you encountered it before.

danielzeljko commented:
>  Also added vscode directory to gitignore, it was annoying for me :D

Can you please elaborate on what _specifically_ was causing annoyance for you with the `.vscode` directory? I'm asking this because I do not see this folder committed to the repository. Also, it's worth noting that our project uses `.editorconfig` file to maintain shared editor settings. These settings should not be altered with personal settings.

If the `.vscode` folder is causing issues for you across multiple projects, then you might want to add it to a global gitignore file.

Here's my `.gitignore_global` for reference:

```
*~
.*.sw[nop]
.DS_Store
.env
__pycache__
env/
venv/
node_modules/
build/
.vscode/
```

> If number of correctors get 0 then the is_corrected field should be 0, right ?

Yes, you got it. In the legacy codebase, there was an issue with filtering and sorting on a bool field in Django 2 and the PG version I was using. This was my hacky solution to address it. This might be something to eventually look into / change back to a bool field as it's highly likely that whatever was causing this issue has been addressed.

> And added the management command as you said, but i don't know how much load is on the database and don't have any estimation of the number of posts that need to be updated, so i just wrote the simplest possible way.

I'll look into this tomorrow with production data

> There's something problem in japanese splitting, it makes the pipline fail, i didn't find how to fix it, maybe you encountered it before.

Ah, that issue makes sense. This is my fault. I'm sorry about that. What's happening is that you're missing the dictionary files. Can you update the README:

```
## Installing Dictionaries

We use [Fugashi](https://github.com/polm/fugashi) and [NLTK](https://www.nltk.org/) for text parsing. Fugashi parses Japanese text and NLTK parses text in various languages.

Fugashi

    $ python -m unidic download
    
NLTK

I typically tend to install all of the datasets and models, but you'll most likely want to download the popular subset of NLTK data.

   $ python -m nltk.downloader popular
```

In order to fix this failing test, we need to update `ci.yml`:

      - name: Setup Fugashi
        run: python -m unidic download

itshesam commented:
@danielzeljko 

> Can you please elaborate on what specifically was causing annoyance for you with the .vscode directory?

I had ```settings.json``` in ```.vscode``` directory, and it was excluded from ignoring by ```!.vscode/settings.json``` in ```.gitignore```. I actually don't know why there is such thing, shouldn't all files in ```.vscode``` be ignored ?
I didnt know about ```.gitignore_global```, it was nice. Anyway i reverted the changes in ```.gitignore```

> Can you update the README:

I updated the ```README``` and ```ci.yml```. I added the instructions as a part of installation in ```README```, because it's presence is necessary  when the code is running, otherwise ```split_sentences``` is going to raise errors on japanese posts.

danielzeljko commented on langcorrect/corrections/tests.py:
Let's follow the same file structure for tests as in our other apps (ie: posts, languages, users, etc) that have tests for consistency.

danielzeljko commented on langcorrect/corrections/tests.py:
```diff
@@ -1,3 +1,55 @@
+
+        return make_corrections(request, post.slug)
+
+    def get_native_speaker_for_post(self, post):
```
`Post` has a model property for this. See `get_correctors`.

danielzeljko commented:
> I actually don't know why there is such thing, shouldn't all files in .vscode be ignored ?

I'm not a senior+ engineer or have corporate experience, so I'm honestly not sure. What I can say is that _it depends_. For example, Mozilla uses `.vscode` to share common tasks and extensions. https://searchfox.org/mozilla-central/source/.vscode.

itshesam commented:
@danielzeljko 

Thanks for the review, it was really nice. The factories for tests was nice. I think i resolved all of the comments except ```Subquery``` thing that you said, that i didn't get what exactly you mean. If you meant that we could query both in just one database query, i doubt if it's possible, but i'll look more into it.
Let me know if i resolved the issues or not.

danielzeljko commented:
@itshesam I see you've made some updates. Can you mark conversations as resolved where applicable? Thanks!

danielzeljko commented:
>  If you meant that we could query both in just one database query, i doubt if it's possible because there are 2 different tables, but i'll look more into it.

It's not that we're querying both tables in a single DB query, but we're using a single `Subquery` to unify the post ids, and using that to update the post objects. This way, we're combining data from both tables in a memory-efficient way.

itshesam commented:
@danielzeljko 

I updated the code. I also created a separate issue(#269) and pull request for downloading Japanese dictionaries.
And i had a question in my mind that why do we have both CorrectedRow and PerfectRow ? i mean couldn't they be merged as one ?
And how can i reach out to you on discord ? can i message you ? it was said to not send private messages without prior approval :D and i also didn't know what is your username

danielzeljko commented:
> And i had a question in my mind that why do we have both CorrectedRow and PerfectRow ? i mean couldn't they be merged as one ?

Great question! Yes, merging the tables with a `category` field is definitely an option. The trade-off is that we'd lose semantic clarity and we'd have to deal with nullable fields. I chose to separate the tables mainly for the sake of clarity, although it does at times make joins and aggregations a bit of a pain in the butt to deal with... There's always a trade-off between normalization and the practical aspects of querying and maintaining the DB. That's what makes DB design tough!

> And how can i reach out to you on discord ? can i message you ? it was said to not send private messages without prior approval :D and i also didn't know what is your username

Please use the dev channels so we can all be in the loop 🙌

itshesam commented:
@danielzeljko 

Is there anything left to be resolved ?

danielzeljko commented on langcorrect/corrections/tests/test_views.py:
```diff
@@ -0,0 +1,49 @@
+                "original_text": post_row.sentence,
+                "corrected_text": updated_sentence,
+                "feedback": "",
+                "type": "",
```
What is `type` referring to here? Is this being used in the view?

Edit: Remove this as well because it's not being used in our view.

danielzeljko commented on langcorrect/corrections/tests/test_views.py:
```diff
@@ -0,0 +1,49 @@
+        corrections_data = [
+            {
+                "sentence_id": post_row.id,
+                "original_text": post_row.sentence,
```
Let's remove `original_text` for consistency and maintainability as it is not being used in our view. We should only pass in data that the view expects to receive.

danielzeljko commented on langcorrect/corrections/tests/test_views.py:
We're going to eventually have more tests in this file. With that in mind, we should refactor this a bit in order to keep the tests more maintainable. Right now we're not keeping things DRY because we will pass the corrector to all of our test functions. It would be a good idea to place all of the common objects in a `setUp` method. This way all of those objects will be available globally throughout our test functions.

Something like this:

```
class TestPostCorrection(TestCase):
    def setUp(self):
        # self.factory = RequestFactory()

        self.language_level = LanguageLevelFactory.create(level=LevelChoices.NATIVE)
        self.language = self.language_level.language
        self.post = PostFactory.create(language=self.language)
        self.corrector = self.language_level.user

        self.client.force_login(self.corrector)


    def perform_correction_request(self, updated_sentence, action):
        pass
        # update this

    def test_deleting_all_corrections_should_unset_iscorrected(self):
        pass
        # update this
```

I commented out `self.factory` because we weren't using it to make requests.

itshesam replied:
Actually I'm not really into this change this way. The test setup should be a general thing that applies to every test, and also i want to see the preparation of objects of a single test in that test so i can easily track what the test is doing.
For example maybe we want to check if a non-logged-in user can change the post or not, so we do not want to use ```force_login``` here. 

PS: Man you are really a perfectionist :)))) Actually i really like that you are trying to make it the most clean way possible, as these conversations provide me with reasons why something might be better over another thing. The only thing is I wish our time gap between our responses would be shorter, so I could get feedback more often.

danielzeljko replied:
It's not that I'm a perfectionist, but I do try to stick with common conventions because they're there for a reason and they serve a valuable purpose. Consistency is key unless we find an approach that's _obviously_ better. If some tests require unique setups, then we should be able to provide overrides as needed.

While _most_ of our tests for corrections will require a logged-in user, you do bring up a fair point about clarity. If test names and docstrings do not offer enough context, then why don't we create a dedicated class for anon tests like `PostCorrectionsAnonTests`? This way we are still keeping our tests DRY without sacrificing the clarity that you mentioned.

I appreciate your eagerness for prompt feedback. I'll try to improve my response time, but please bear in mind that I'm not always on LC full-time, especially on the weekends. So, I will not be able to respond until the following Monday. There's not much we can do about our time gap, but as we discussed on Discord, sticking to the bullet points will greatly streamline the review process. 

Regarding #181, I am aware of its significance and will review it thoroughly in due time. Given its hold on subscriber data, we'll approach it with the caution it deserves.

itshesam replied:
I'm not sure if I got your idea on anon tests correctly or not. But there is no actual benefit to that and it's somehow misleading as for someone new, it wouldn't be obvious at all why there are 2 separate classes for a single reason. Moreover I don't think dividing tests based on the criterion that they use the same data or not would be a good idea. I think it's a desperate solution to achieve DRY at any cost. As a result the change itself doesn't have any meaning but is a means to achieve a desired goal exclusively.

danielzeljko reviewed (changes requested):
Sorry for getting to this late. I've requested some changes.
